homeassistant/components/abode/config_flow.py

```diff
@@ -26,48 +27,151 @@ def __init__(self):
             vol.Required(CONF_PASSWORD): str,
         }
 
+        self.mfa_data_schema = {
+            vol.Required(CONF_MFA): str,
+        }
+
+        self._username = None
+        self._password = None
+        self._polling = None
+        self._cache = None
+
     async def async_step_user(self, user_input=None):
         """Handle a flow initialized by the user."""
         if self._async_current_entries():
             return self.async_abort(reason="single_instance_allowed")
 
         if not user_input:
-            return self._show_form()
+            return self.async_show_form(
+                step_id="user", data_schema=vol.Schema(self.data_schema)
+            )
+
+        self._username = user_input[CONF_USERNAME]
+        self._password = user_input[CONF_PASSWORD]
+        self._polling = user_input.get(CONF_POLLING, False)
 
-        username = user_input[CONF_USERNAME]
-        password = user_input[CONF_PASSWORD]
-        polling = user_input.get(CONF_POLLING, False)
-        cache = self.hass.config.path(DEFAULT_CACHEDB)
+        return await self.async_step_login(step_id="user")
+
+    async def async_step_login(self, step_id):
+        """Handle Abode login."""
+        self._cache = self.hass.config.path(DEFAULT_CACHEDB)
+        errors = {}
 
         try:
+            abode = Abode(
+                auto_login=False,
+                get_devices=False,
+                get_automations=False,
+                cache_path=self._cache,
+            )
             await self.hass.async_add_executor_job(
-                Abode, username, password, True, True, True, cache
+                abode.login, self._username, self._password
             )
 
         except (AbodeException, ConnectTimeout, HTTPError) as ex:
+            if ex.errcode == MFA_CODE_REQUIRED[0]:
+                return await self.async_step_mfa()
+
             LOGGER.error("Unable to connect to Abode: %s", str(ex))
+
             if ex.errcode == HTTP_BAD_REQUEST:
-                return self._show_form({"base": "invalid_auth"})
-            return self._show_form({"base": "cannot_connect"})
-
-        return self.async_create_entry(
-            title=user_input[CONF_USERNAME],
-            data={
-                CONF_USERNAME: username,
-                CONF_PASSWORD: password,
-                CONF_POLLING: polling,
-            },
+                errors = {"base": "invalid_auth"}
+            else:
+                errors = {"base": "cannot_connect"}
+
+        if errors:
+            return self.async_show_form(
+                step_id=step_id, data_schema=vol.Schema(self.data_schema), errors=errors
```

Comment:
I think once I refactor this to be a helper method based on the feedback above, that should resolve this issue since this isn't actually a step but really a helper method.